Head shape

nchataing opened the pull request:
a comment

Commit:
Add head_shape type

Commit:
Adapt datarepr to generate head_shape thunks in unboxed constructors

Commit:
Add dump option for head shapes

Commit:
Add tests

Commit:
sw_num{non}consts fix

Commit:
Add head_shape computing algorithm

gasche commented on typing/types.ml:
```diff
@@ -417,18 +417,38 @@ type constructor_description =
+and unboxed_cache =
+  { cache_ty: type_expr;
+    cache_shape: head_shape option ref;
+  }
```
I think that `unboxed_ty` and `unboxed_head_shape` would read better as field names?

gasche replied:
(Then maybe `unboxed_cache` could become `unboxed_data`)

nchataing replied:
I prefer having `unboxed_shape` then (`unboxed_head_shape` is a very long name).

gasche commented on typing/types.ml:
```diff
@@ -417,18 +417,38 @@ type constructor_description =
+  | Cstr_unboxed c1, Cstr_unboxed c2 ->
+      (* Possible tags of different unboxed constructors are disjoint,
+         so in particular their types must be different. *)
+      c1.cache_ty == c2.cache_ty
```
This is wrong in the case of an empty type.

```ocaml
type empty = |
type t = Foo of empty [@unboxed] | Bar of empty [@unboxed]
```

We could already use your suggestion of checking for physical equality of the `unboxed_head_shape` references.

gasche commented on typing/datarepr.ml:
```diff
@@ -108,17 +108,15 @@ let constructor_descrs ~current_unit ty_path decl cstrs rep =
+          | _, Cstr_tuple [] -> (Cstr_constant idx_const,
+                   describe_constructors (idx_const+1) idx_nonconst rem)
+          | _, _ -> (Cstr_block idx_nonconst,
+                   describe_constructors idx_const (idx_nonconst+1) rem) in
```
It would be nicer to have a clean failure if using `[@unboxed]` on a type that cannot be unboxed because it is not a one-parameter variant. I think that we should cover all `true` cases explicitly, instead of having some of them default to the standard (not unboxed) case.

Note: The standard way to raise an error here would be to use an `error` type and an `Error of Location.t * error` exception, with a registered error printer. (Typedecl does this already and can be looked at for inspiration.) An easier way that is nicer than the current code would be to use `Location.errorf` with a locally-provided error message; this is not the best approach, but it could be a good compromise if you prefer to do the `error` boilerplate only later.

gasche commented on typing/typedecl_unboxed.ml:
```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+        else if Path.same p Predef.path_nativeint
+             || Path.same p Predef.path_int32
+             || Path.same p Predef.path_int64 then
+          block_shape [Obj.custom_tag]
```
I think that we could replace this sea of `if Path.same` tests with a function defined outside the Head_shape logic:

```ocaml
type primitive_type =
  | Float
  | String
  | Array of type_expr
  | ...

val match_primitive_type : Path.t -> primitive_type option
```

gasche commented on typing/typedecl_unboxed.ml:
```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+  and of_typedescr env ty_descr ty_decl ~args
+                   callstack_map new_callstack =
+    (* we called scrape_ty on the type expression in the function above, thus
+       it isnt a single constructor unboxed type.
```
This comment is now outdated, we don't use `scrape_ty` anymore

gasche commented on typing/typedecl_unboxed.ml:
```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+        of_type_with_params ty
+    | Type_record (_, Record_unboxed _)
+    | Type_record (_, Record_inlined _)
+    | Type_record (_, Record_extension _) -> assert false
```
I think that `failwith "TODO"` would be clearer if this is what we mean here. My understanding is that `Record_unboxed` is actually an `assert false`, as all possible cases where handled above. For this I usually prefer using nested matching

```ocaml
Type_record (fields, Record_unboxed _) ->
  begin match fields with
  | [{lbl_arg = ty; _}] -> ...
  | _ -> assert false
  end
```

gasche commented on typing/typedecl_unboxed.ml:
```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+    | Type_record (_, Record_inlined _)
+    | Type_record (_, Record_extension _) -> assert false
+    | Type_open -> block_shape [0]
+    | Type_variant ([],_) -> none
```
reading this code again, I find `none` a bit cryptic here; maybe `empty` would read better, even if the symmetry with `any` is broken?

nchataing replied:
Maybe renaming both `any` to `any_shape` and `none` to `empty_shape` would read better.

gasche replied:
In general they will be used as `Head_shape.{any,none,empty}`, so it will read well. We happen to be inside the module so it's a bit more demanding on the naming.

nchataing replied:
I think those values won't be in the signature of `Head_shape` so only to be used in the implementation.

gasche commented on typing/typedecl_unboxed.ml:
```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+    | Type_variant ([],_) -> none
+    | Type_variant ((fst_descr :: _) as cstr_descrs,_) ->
+        (* the head shape of boxed constructors is equivalent to the nb of
+           constant constructors and the nb of non constant constructors *)
```
reading this comment after the fact, I first thought that it meant that this `Type_variant ..` case can only be reached by types without unboxed constructors. Maybe a first sentence saying something like "We compute the shape of all boxed constructors, and then the shapes of each unboxed constructor." would help.

gasche commented on typing/typedecl_unboxed.ml:
```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+           base head shape of boxed constructors *)
+        List.fold_left disjoint_union boxed_shape unboxed_shapes
+
+  let fill_cache env {cache_ty; cache_shape} =
```
The name is correct, but `get`` would also work and may read better at the callsite.

(Or: have a `fill_cache` version that returns `unit`, and a `get` version that fills the cache and returns the head shape. `fill_cache head_shape_data` reads better than `ignore (get head_shape_data)`.)

gasche reviewed (commented):
This is a first batch of comments, looking at the commits in order (I didn't get to the pattern-matching part yet). The split / history cleanup is very nice, thanks!

gasche commented on typing/typedecl_unboxed.ml:
```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+
+  let rec of_type env ty callstack_map =
+    let ty = Ctype.repr ty in
+    check_annotated ty callstack_map;
```
Note: I still wonder if we could use `Ctype.expand_head_opt` here. (Maybe we didn't try well enough last time?). I think this will be easier to answer (and discuss with more knowledgeable people) in the `get_unboxed_type_repr` PR, assuming that PR gets a better testsuite than we have now. (I have `-dtypedecl-unboxed-repr` somewhere on my TODO list towards that.)

gasche commented on lambda/matching.ml:
```diff
@@ -2734,20 +2734,81 @@ let combine_constant loc arg cst partial ctx def
+              | None ->
+                  invalid_arg "Matching.split_cases: missing head shape"
+              | Some shape -> shape
+            in
```
This could be just `Head_shape.get head_shape_data`.

gasche commented on lambda/matching.ml:
```diff
@@ -2819,48 +2880,92 @@ let combine_constructor loc arg pat_env cstr partial ctx def
+               | Some act -> act
+               | None ->
+                  call_switcher loc fail_opt arg 0 (n - 1) cases.consts
+               end
```
nitpick: it's weird that we match on `any_const` in the record pattern, but access `cases.consts` by projection. I would use the same approach for both.

gasche commented on lambda/matching.ml:
```diff
@@ -2819,48 +2880,92 @@ let combine_constructor loc arg pat_env cstr partial ctx def
+                       -- given that split_cases will expand several-tag
+                       constructors into several list items. *)
+                      if List.length cases = complete_case_count then
+                        same_actions cases
```
As far as I can tell, this `FIXME` is left over in the complete PR, and this `TODO` is not handled (and, if we wanted to follow its suggestion, would require adding yet another field to the Datarepr stuff). What is your plan for those two?

gasche commented on lambda/matching.ml:
```diff
@@ -2819,48 +2878,109 @@ let combine_constructor loc arg pat_env cstr partial ctx def
+                      in match bounds with
+                        | Max imm, Max tag -> imm, tag
+                        | Unbounded, _ | _, Unbounded ->
+                            invalid_arg "Matching.combine_constructors"
```
We know that we have `single_nonconst_act = None` here, so unless I am mistaken the `_, Unbounded` case can never happen (it could be turn into an `assert false` with a comment explaining why it is unreachable.) On the other hand, why is the `Unbounded, _` case impossible? I would think that it is possible with something like:

```ocaml
type t = Int of int [@unboxed] | Cstr1 of unit | Cstr2 of unit`

let test = function
  | Int n -> n
  | Cstr1 -> 1
  | Cstr2 -> 2
```

If we have `Shape_any` on immediate values, we should generate a switch with `sw_numconsts = 0`, which should work well with our definition of `fail_opt` above (it will in fact go to the `any_const` action on immediates). This is not what the code below is doing, so I would guess we have a bug? (But then I have not tested anything.)

gasche commented on lambda/translcore.ml:
```diff
@@ -333,7 +334,7 @@ and transl_exp0 ~in_new_scope ~scopes e =
         Cstr_constant n ->
           Lconst(const_int n)
-      | Cstr_unboxed ->
+      | Cstr_unboxed _ ->
```
Was this not forgotten in older commits that changed the type of `Cstr_unboxed`?

nchataing replied:
This bit of code looks correct as far as I know (we don't need the head_shape here).

gasche replied:
If I understand correctly, you had a first commit that added arguments to the `Cstr_unboxed` constructor, and then this change came in a later commit, when it should have been part of this first commit. (Maybe I was mistaken when I made this remark, because the code of the first commit would not compile if this was true.)

nchataing replied:
It is indeed introduced in the first commit of the branch

gasche commented on toplevel/genprintval.ml:
```diff
@@ -388,7 +388,8 @@ module Make(O : OBJ)(EVP : EVALPATH with type valu = O.t) = struct
+                        assert false
                       else if O.is_block obj
                       then Cstr_block(O.tag obj)
                       else Cstr_constant(O.obj obj) in
```
This genprintval.ml adaptation is tricky and will require more work. What the code is doing is to use type information to print a "raw value" in the toplevel. For boxed constructors, we use `Obj.tag` to get the tag at runtime, and then we query the constructor-descriptions to get the expected arity and type of the argument(s).

What should we do with unboxed constructor? What we need in general is a mapping from the tag (or immediate value) of the runtime value `obj` to one specific constructor, and to do this we need to inspect head shapes: if the value comes from an unboxed constructor, we can find it by comparing the runtime tag with its head shape.

Note: this does not work well for unboxed constructor whose argument is an empty type, if we allow several such constructors. But this is not a problem because such unboxed constructors are un-inhabited, the runtime value we have will never come from one of those (non-disjoint) unboxed constructors anyway.

gasche replied:
Discussing this again: things would work correctly if `Datarepr.find_constr_by_tag` was aware that the "raw tag" or "runtime tag" passed (maybe we should use a type different from `constructor_tag`; `runtime_constructor_tag`, without an `unboxed` case?) may not be one of the boxed tags, but it may also belong to the head shape of (at most one of) the unboxed constructors.

(We may need to force the `head_shape` here, instead of within `Datarepr`.)

gasche replied:
@nchataing remarks that this should be computed by matching not on the variant type declaration, but on the type description.

gasche reviewed (commented):
More comments.

Commit:
WIP: restrict matching test to ocamlc.byte-only

gasche commented on typing/typedecl_unboxed.ml:
```diff
@@ -307,16 +306,39 @@ module Head_shape = struct
+      | Tconstr (p, _, _) -> p
+      | _ -> assert false (* XXX *)
+    in
+    let variant_decl = Env.find_type variant_path env in
```
I propose to factorize this logic as a helper functions in Types:

```ocaml
val cstr_type_path : constructor_description -> Path.t
```

moving the `assert false` to one central place instead of duplicating it at each use-site of this logic.

gasche replied:
Note: this could/should be a "refactoring PR" against trunk, also improving many current use-sites for `cstr_res` that are doing exactly this. Would you like to send it, or should I?

nchataing replied:
I'll do it :-)

nchataing replied:
The only problem I see with this refactoring is that the "error handling" is not the same at each use site of `cstr_res`.

And if we return an option type, it doesnt really solve the problem at hand

nchataing replied:
`mark_constructor_description_used` suggests that it is sound to just `assert false`

gasche commented on typing/types.ml:
```diff
@@ -415,6 +413,17 @@ type constructor_description =
+    vd_unboxed: int;                    (* Number of unboxed constructors *)
+    mutable vd_max_values: (int bound * int bound) option;
+    (* Maximum immediate value and block tags taking into account unboxed
+       constructors (only available after head-shape analysis) *)
```
I'm not convinced by the use of `(int bound * int bound)` here:
- I think it makes the type `max_val_of_cstr_descr` harder to understand (than with, say, a record type with named field)
- is the plan to store all types that are only available later in a single tuple? we discuss that we also need to represent something like `vd_unboxed_consts` and `vd_unboxed_nonconsts, will they also go in this same `mutable ... : ... option` field?

I think that these two issues would be improved by going back to a single field per information, all of them `mutable` with option, but with a helper function for each field in Typedecl

```ocaml
(* types.ml *)
  mutable vd_max_const_tag : int option;

(* typedecl.ml *)
val cstr_max_const_tag : Env.t -> Types.constructor_description -> int
```

that takes care of recomputation and caching.

(I suggest defining the function in `Typedecl` rather than `Typedecl_unboxed` because the user cannot really guess that this logic is related to unboxed constructors.)

Implementation sketch (untested):
```ocaml
(* in typedecl.ml *)
val get_head_shape : Env.t -> Path.t -> head_shape
...

let head_shape_bound = function
  | Shape_any -> Unbounded
  | Shape_set li -> Max (List.fold_left max 0 li)

let get_variant_data env cstr_desc =
  let vd = cstr_desc.variants in
  match vd.vd_max_const_tag, vd.vd_max_nonconst_tag with
  | Some _, Some _ -> vd
  | None, _ | _, None ->
    if vd.vd_unboxed = 0 then begin
      vd.vd_max_const_tag <- Max (vd.vd_consts - 1);
      vd.vd_max_nonconst_tag <- Max (vd.vd_nonconsts - 1);
    end else begin
      let type_path = Types.cstr_type_path cstr_desc in
      let shape = get_head_shape env type_path in
      vd.vd_max_const_tag <- head_shape_bound shape.shape_imm;
      vd.vd_max_nonconst_tag <- head_shape_bound shape.shape_blocks;
  end;
  vd

let cstr_max_const_tag env cstr_descr =
  (get_variant_data env cstr_descr).vd_max_const_tag

let cstr_max_nonconst_tag env cstr_descr =
  (get_variant_data env cstr_descr).vd_max_nonconst_tag
```

gasche reviewed (commented):
The end of my review (for `head_shape` as it was two days ago, if I understand correctly).

nchataing committed:
Variant data refactoring

gasche commented on lambda/matching.ml:
```diff
@@ -2908,27 +2908,24 @@ let combine_constructor loc arg pat_env cstr partial ctx def
+                          same_actions cases
+                        else
+                          None
+                      | None -> assert false
```
It took me a while to understand what was going on here, as I had forgotten about the context. There should be a comment that explains why we will not reach `None` here. (I think that the reason is that `complete_case_count` must be `Some` if `any_case` is `None`.)

gasche commented on lambda/matching.ml:
```diff
@@ -2959,27 +2956,20 @@ let combine_constructor loc arg pat_env cstr partial ctx def
+                    let sw_numblocks =
+                      match Typedecl.cstr_max_block_tag pat_env cstr with
+                      | Some n -> n + 1
+                      | None -> invalid_arg "Matching.combine_constructors"
```
There should be a comment that explains why we `invalid_arg` here. If I follow correctly, we already checked that `single_nonconst_act` is `None` so we must have at least two non-constant actions. (So in fact, unless I'm missing something, we know statically that this `invalid_arg` never raises so we could even use `assert false`.)

gasche commented on typing/datarepr.ml:
```diff
@@ -101,20 +101,13 @@ let constructor_descrs ~current_unit ty_path decl cstrs rep =
   let ty_res = newgenconstr ty_path decl.type_params in
   let variant_unboxed = Builtin_attributes.has_unboxed decl.type_attributes in
   let num_consts = ref 0 and num_nonconsts = ref 0 and num_unboxed = ref 0 in
+  let variant_data = ref None in
```
The declaration placed here gives the impression that this reference will also be mutated by the constructor-traversal below. I would just use `variant_data = ref None;` in the record creation at the end.

nchataing replied:
Here we would like to have the `variant_data` shared between constructors (so that we don't recompute it for each constructor of the variant). Then we could just move the let-binding under the `List.iter` just below.

gasche replied:
Ah, of course... Yes, moving it below would be nice. And please include a comment on the sharing: "Gabriel out of touch" is a proxy for the people unfamiliar with the code having comprehension troubles.

gasche commented on typing/typedecl.ml:
```diff
@@ -1924,3 +1924,36 @@ let () =
+    - val get_head_shape : Env.t -> Path.t -> head_shape;
+    - revise the Head_shape module signature;
+    - write the get_variant_head_shape function;
+    - write helper function to get the desired variables;
```
I think this TODO item is now stale as all points have been performed?

gasche replied:
(The first point has not been implemented, but with the new `Btype` helper function it's easy to do it at the callsite. You decided that we don't need it anymore, and I think it's reasonable.)

gasche commented on lambda/matching.ml:
```diff
@@ -2959,27 +2956,20 @@ let combine_constructor loc arg pat_env cstr partial ctx def
                     in
                     let sw =
-                      { sw_numconsts = numconsts + 1;
+                      { sw_numconsts;
```
It's easy to be confused by the subtle yet important difference between `numconsts` and `sw_numconsts`, `numnonconsts` and `sw_numblocks`. (Maybe `numblocks` is a nicer name than `numnonconsts` by the way?)

I think we should have comments on the binding sites of `num{consts,nonconsts}`, `complete_case_count` and `sw_num*` that explain what those values mean and what they are used for. This would be very helpful to people less familiar with the code.

gasche commented on lambda/matching.ml:
```diff
@@ -2959,27 +2956,20 @@ let combine_constructor loc arg pat_env cstr partial ctx def
+                    let sw_numconsts =
+                      match Typedecl.cstr_max_imm_value pat_env cstr with
+                      | Some n -> n + 1
+                      | None -> 1 (* XXX *)
```
What do we want here? `max_imm_value` is None when we have a `Shape_any`. In this case the `any_const` action is integrated into the `fail_opt` action above, so I would expect that we have *no* constant actions in the switch, suggesting `sw_numconsts = 0`.

gasche replied:
We discussed this. The problem is that  `simplif.ml` assumes that if there is a fail-action, then one of the two cases is partial. With this code the assumption is wrong when `fail_opt = None` but `any_const` is not (so the "merged" fail_opt is Some), and we cover all nonconst cases.

So in theory we should put `max_int` here if there is an `any_const`, but this would blow up compilation later that creates an array of this size.

`inline_lazy_force_switch` has the same issue, there solution is to do a `Pisint` test first and then a `switch` (with `sw_numconsts = 0`, `consts = []`) only in the block case. We can do the same here:
- if `any_const = None`, compile as in the current PR
- if `fail_opt = None` and `any_const = Some`, instead of defining a new `fail_opt` we should first generate a `Pisint` test, send to the `any_const` action in the then branch, and generate a switch only in the else branch
- if `fail_opt = Some` and `any_const = Some`, we can either generate a Pisint first (as above) or generate code as in the current PR (probably generate nicer code).

The solution of using `sw_numconsts = 1` may be unsafe (we are not sure) if the compiler then assumes that the immediate input may only be `0`, not any immediate.

gasche reviewed (commented):
Thanks, the refactoring looks good.

I think that we are at the point where we could consider a "draft PR" against the main repository. (After cleaning up the history.) There are three remaining TODOs:
- adding some comments (I also asked for some extra comments in my last review): you should do this before submitting upstream
- handling of the `Record_extension` case: no strong opinion, could be done before or after
- trying the `Ctype.expand_opt` version: I think we can keep it for after if you prefer

nchataing committed:
Move variant_data binding (and add comment)

nchataing committed:
replace invalid arg by an assert false

gasche committed:
document typing/types.mli + renamings

gasche committed:
document typedecl_unboxed.ml{,i}

gasche committed:
document changes in lambda/matching.ml

gasche committed:
document our disappointing bytecode generation

gasche committed:
a TODO item with a concrete proposal for more compact bytecode

nchataing committed:
Compile switch with a Pisint if the any_const case action is Some.

gasche committed:
refactoring commit

gasche committed:
Revert "refactoring commit"

This reverts commit 56dd15e1018883b6a9423a9265185402d4fcb37d.

I am not sure about this commit, because it causes the compiler to
generate worse bytecode in some cases: if any_const is None but
single_const_action is Some, then we will generate a test followed by
a switch (in the nonconst case) instead of a single switch.

gasche committed:
RAEDME.head_shape.md: prepare a description for PR submission

gasche committed:
PR description: describe the implementation

gasche committed:
spec: why we are not turning imm/tag into constructed types

gasche committed:
[minor] typedecl_unboxed: refactor into Head_shape.of_primitive_type

gasche committed:
[minor] matching.ml: sum/product refactoring for Cases

gasche committed:
update README.head_shape

gasche committed:
split the head-shape spec in two files, specification (-> RFC) and implementation (-> PR)

gasche committed:
rephrase the separability proposal

nchataing committed:
Update HEAD_SHAPE.spec.md

gasche committed:
[minor] more code markers

gasche committed:
slightly more in .impl.md

gasche committed:
mention separability in the spec

gasche committed:
HEAD_SHAPE.impl.md: add genprintval handling in the TODO

gasche committed:
[minor] genprintval.ml: helper functions to print variant/record values

gasche committed:
[minor] genprintval.ml: clarify exception flow

gasche committed:
[WIP] Typedecl_unboxed: new operations on head shapes

We introduce explicitly the "head" of a value; head shapes are finite
approximations of sets of heads. We can compute the head of a value
and check whether a head is included in a given head shape. This will
be useful for toplevel/genprintval.ml which must dynamically decide
the (possibly unboxed) constructor of a dynamic value of known type.

TODO: we should discuss the placement of the types and operations, and
document them;
- maybe move the type of heads in types.mli
- or maybe it's time to introduce dedicated compilation units,
  Head_shape_types (on which Types depends) and Head_shape (on which
  other units depend)?

gasche committed:
head_shape: expose printer for debugging

gasche committed:
toplevel/genprintval.ml: support unboxed constructors

gasche committed:
[refactoring] Outcometree: introduce a record type for constructors

gasche committed:
outcometree: print [@unboxed] attributes

gasche committed:
oops: restore the fatal error when declaring conflicting types

(report from Guillaume Munch-Maccagnoni)

gasche committed:
more precise head shape for function types

(suggestion from Guillaume Munch-Maccagnoni)

nchataing committed:
separability

nchataing committed:
fix bug with global [@@unboxed] annotation

gasche committed:
update the testsuite

gasche committed:
head_shape: change separability computation (still unfinished)

nchataing commented:
I see, my error was that I was thinking `any_shape` would be not separated by default, ruling out declarations like `type t = A of abstr [@unboxed]`, but it makes sense to be separated.

gasche committed:
head-shape spec: genprintval is done
